Display PersistentVolumeSize in `crc status`

vyasgun opened the pull request:
Enhance the crc status command to display the PersistentVolumeSize from the CRC configuration.

**Fixes:** Issue #4191 

## Solution/Idea
The `/status` handler should return fields specifying the PVC size and usage (allocated storage).

## Proposed changes

 - Fetch and display the PersistentVolumeSize value in the crc status output.
 - Add new fields to the ClusterStatus and CrcStatus structs to support this feature.
 - Update the CRC daemon to include these changes.

## Testing

Run `crc status`
```
gvyas-mac:~ gvyas$ crc status
CRC VM:                  Running
MicroShift:              Running (v4.15.12)
RAM Usage:               1.62GB of 3.904GB
Disk Usage:              6.824GB of 20.41GB (Inside the CRC VM)
Persistent Volume Usage: 10.67GB of 16GB (Allocated)
Cache Usage:             74.36GB
Cache Directory:         /Users/gvyas/.crc/cache
```

vyasgun commented:
/retest

anjannath commented:
it is showing the value of the `persistent-volume-size` config option, but it'd be more useful if the usage was shown like the `Disk Usage` (how much is used and how much is free)

e.g
```
CRC VM:                  Running
OpenShift:               Running (v4.16.0)
RAM Usage:               6.374GB of 10.92GB
Disk Usage:              22.86GB of 32.68GB (Inside the CRC VM)
Persistent Volume Usage: 7GB of 15GB
Cache Usage:             64.58GB
Cache Directory:         /home/anjan/.crc/cache
```

----

also this should be shown only when using the `microshift` preset for now, since for `openshift` preset the config doesn't have any effect yet

```
% crc config get preset
Configuration property 'preset' is not set. Default value 'openshift' is used

% crc status
CRC VM:                 Running
OpenShift:              Running (v4.16.0)
RAM Usage:              6.374GB of 10.92GB
Disk Usage:             22.86GB of 32.68GB (Inside the CRC VM)
Persistent Volume Size: 15GB
Cache Usage:            64.58GB
Cache Directory:        /home/anjan/.crc/cache
```

praveenkumar commented on pkg/crc/api/client/types.go:
```diff
@@ -19,14 +19,16 @@ type StartResult struct {
+	RAMUse               int64
+	RAMSize              int64
+	PersistentVolumeUse  int
+	PersistentVolumeSize int
```
doesn't that mean it will be required field for openshift preset also which is not the case right? So may be use `json:"PersistentVolumeUse/Size,omitempty"` for this also

vyasgun replied:
Done

vyasgun committed:
Display PersistentVolumeUsage in `crc status`

- Enhance the crc status command to display the PersistentVolumeUsage and PersistentVolumeSize from the CRC VM.
- Fetch and display the persistent volume usage value in the crc status output.
- Add new fields to the ClusterStatus and CrcStatus structs to support this feature.
- Update the CRC daemon to include these changes.

vyasgun commented:
@anjannath Just updated the code. Please take a look again, thanks!

praveenkumar commented on pkg/crc/cluster/cluster.go:
```diff
@@ -130,6 +130,29 @@ func GetCPUUsage(sshRunner *ssh.Runner) ([]int64, error) {
+    exit 0
+fi
+sudo df -B1 --output=used $mountpoints | awk ' { sum += $1 } END { printf "%d", sum} '
+`
```
Did you check `$ sudo pvs -o pv_free --noheadings --nosuffix --units k` this provide the free space by default instead of all those grepping.

vyasgun replied:
yes, but it does not give the exact usage of mounted volumes.
On my system, the output is:
```
sudo pvs -o pv_free --noheadings --nosuffix --units k
  9453568.00
```
provides 9.4G of free space which means `15-9.4 = 5.6G` of used space.
```
[core@api ~]$ cat print_pv_usage.sh
#!/bin/bash
mountpoints=$(lsblk --output=mountpoints | grep pvc | tr '\n' ' ')
if [ -z "$mountpoints" ]; then
    exit 0
fi
sudo df -B1 --output=used $mountpoints | awk ' { sum += $1 } END { printf "%d", sum} '

[core@api ~]$ ./print_pv_usage.sh
111681536
```
which is `111.7MB` of used space

cfergeau replied:
And which one is correct? Do you have 111MB of data in your PVs, or 5,6G of data? Or another value?

vyasgun replied:
111MB is the correct one. The way I am fetching this value is as follows:

I have two PVCs:
```
gvyas-mac:~ gvyas$ kubectl get pvc
NAME       STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS          AGE
my-pvc     Bound    pvc-57f25d94-68de-4e76-99a6-983c488d8ba8   2Gi        RWO            topolvm-provisioner   5d8h
my-pvc-1   Bound    pvc-ba22fa21-5f01-4496-9044-3e0a8f8a8cd0   4Gi        RWO            topolvm-provisioner   5d8h
```
Inside the VM, I am using lsblk to find the relevant mountpoints.
```
[core@api ~]$ lsblk
NAME                                              MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
vda                                               252:0    0   31G  0 disk
├─vda1                                            252:1    0  200M  0 part /boot/efi
├─vda2                                            252:2    0  800M  0 part /boot
├─vda3                                            252:3    0    1M  0 part
├─vda4                                            252:4    0    1M  0 part
└─vda5                                            252:5    0   30G  0 part
  ├─rhel-root                                     253:0    0   15G  0 lvm  /var/lib/containers/storage/overlay
  │                                                                        /var
  │                                                                        /sysroot/ostree/deploy/rhel/var
  │                                                                        /usr
  │                                                                        /etc
  │                                                                        /
  │                                                                        /sysroot
  ├─rhel-0ad559a7--8b01--4ed0--b460--5aae3f0fa7c8 253:1    0    4G  0 lvm  /var/lib/kubelet/pods/a1473da7-14fb-4aeb-99fc-f411c97aec1c/volumes/kubernetes.io~csi/pvc-ba22fa21-5f01-4496-9044-3e0a8f8a8cd0/mount
  └─rhel-4be6fea3--c245--42f3--b74a--95d4cf7c4e0a 253:2    0    2G  0 lvm  /var/lib/kubelet/pods/e57ecde7-20dd-4ce6-8180-2b837dc0bffd/volumes/kubernetes.io~csi/pvc-57f25d94-68de-4e76-99a6-983c488d8ba8/mount
```
The bottom two mountpoints correspond to the two PVCs on the cluster. Then I am using `df` to get the usage.

```
[core@api ~]$ sudo df -h /var/lib/kubelet/pods/a1473da7-14fb-4aeb-99fc-f411c97aec1c/volumes/kubernetes.io~csi/pvc-ba22fa21-5f01-4496-9044-3e0a8f8a8cd0/mount /var/lib/kubelet/pods/e57ecde7-20dd-4ce6-8180-2b837dc0bffd/volumes/kubernetes.io~csi/pvc-57f25d94-68de-4e76-99a6-983c488d8ba8/mount
Filesystem                                         Size  Used Avail Use% Mounted on
/dev/topolvm/0ad559a7-8b01-4ed0-b460-5aae3f0fa7c8  4.0G   61M  3.9G   2% /var/lib/kubelet/pods/a1473da7-14fb-4aeb-99fc-f411c97aec1c/volumes/kubernetes.io~csi/pvc-ba22fa21-5f01-4496-9044-3e0a8f8a8cd0/mount
/dev/topolvm/4be6fea3-c245-42f3-b74a-95d4cf7c4e0a  2.0G   47M  1.9G   3% /var/lib/kubelet/pods/e57ecde7-20dd-4ce6-8180-2b837dc0bffd/volumes/kubernetes.io~csi/pvc-57f25d94-68de-4e76-99a6-983c488d8ba8/mount
```

After writing 100MB of data to `my-pvc`:
```
gvyas-mac:~ gvyas$ crc status
CRC VM:                  Running
MicroShift:              Running (v4.15.12)
RAM Usage:               2.037GB of 3.904GB
Disk Usage:              6.778GB of 16.1GB (Inside the CRC VM)
Persistent Volume Usage: 216.5MB of 15GB
Cache Usage:             74.36GB
Cache Directory:         /Users/gvyas/.crc/cache
```

Inside the VM:
```
[core@api ~]$ sudo df -h /var/lib/kubelet/pods/a1473da7-14fb-4aeb-99fc-f411c97aec1c/volumes/kubernetes.io~csi/pvc-ba22fa21-5f01-4496-9044-3e0a8f8a8cd0/mount /var/lib/kubelet/pods/e57ecde7-20dd-4ce6-8180-2b837dc0bffd/volumes/kubernetes.io~csi/pvc-57f25d94-68de-4e76-99a6-983c488d8ba8/mount
Filesystem                                         Size  Used Avail Use% Mounted on
/dev/topolvm/0ad559a7-8b01-4ed0-b460-5aae3f0fa7c8  4.0G   61M  3.9G   2% /var/lib/kubelet/pods/a1473da7-14fb-4aeb-99fc-f411c97aec1c/volumes/kubernetes.io~csi/pvc-ba22fa21-5f01-4496-9044-3e0a8f8a8cd0/mount
/dev/topolvm/4be6fea3-c245-42f3-b74a-95d4cf7c4e0a  2.0G  147M  1.8G   8% /var/lib/kubelet/pods/e57ecde7-20dd-4ce6-8180-2b837dc0bffd/volumes/kubernetes.io~csi/pvc-57f25d94-68de-4e76-99a6-983c488d8ba8/mount
[core@api ~]$ ./print_pv_usage.sh
216539136
```

cfergeau replied:
Thanks for the detailed answer :) 
Seeing the output of `kubectl get pvc`, it looks like the command Praveen gave shows you the claimed space in the PVs, while your command gives you the space actually used on disk. Not sure which one is more important to show.
The PV claims will (I think) limit us, for exmpale if you claimed 15GB of disk space, and you only have 15GB on the partition for the PVs, then you won't be able to do anything more even if you have only used 100MB in the claimed storage.
So maybe it's just the "non claimed" space we want to report, regardless of whether the claimed space is free on not free?

(disclaimer, I'm not that familiar with PVs/..., I hope I did not get the basics wrong in the explanation above :)

praveenkumar replied:
Yes, that's right. So If I create a PV of 2GB then I have only 13G quota left for creating other PV's even my PV is only using the 10MB from that 2GB so I think we should show the non claimed space to user so they can make informed choice about size of a PV when creating one.

vyasgun replied:
Yes, it is true that provisioning PVCs in the future will use information about the claimed space rather than the used space. However, users can easily check this by using `kubectl get pvc`, which shows the capacity of all created PVCs. I thought the purpose of showing this information in `crc status` is to provide VM resource stats similar to disk usage.

praveenkumar replied:
> is to provide VM resource stats similar to disk usage.

Disk usage is correct info for user and that space can be consumed but in case of pvc it provide a false info to the user since this space is only use for PV creation so user should know how many he can create using a informed decision about how much space on the disk for pv creation.

vyasgun replied:
Another point to consider is overprovisioning. If I use a storage class with overprovisioning ratio set to greater than 1, I can allocate more space than the configured storage. For instance, in the example below, 18Gi was allocated in total to the two PVCs even though the configured upper limit is 16Gi.
```
gvyas-mac:~ gvyas$ crc config get persistent-volume-size
persistent-volume-size : 16

gvyas-mac:~ gvyas$ kubectl get pvc
NAME       STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS       AGE
my-pvc-1   Bound    pvc-4d9f8015-ce4c-4d1c-8c4a-e4e5622d5489   10Gi       RWO            thin-provisioner   12s
my-pvc-2   Bound    pvc-584e7136-d8ea-49ac-8569-3c8e1f47613d   8Gi        RWO            thin-provisioner   4m55s
```

anjannath replied:
i think we should show the allocated space and not the actual used space, as it'd be complicated if we want to show actual space used if user has define their own SC with higher over provisioning ratio and only allocated space when over provisioning is 1

maybe we can add `6.778GB of 16.1GB (Allocated)` to hint to the user that its allocated space and not used space!

praveenkumar replied:
> Another point to consider is overprovisioning

I would discourage that since if that pod really consumed that much space then it lead to data corruption so for our end better to tell user how much storage PV used during creation and how much left to allocate for the user.

vyasgun replied:
Please check the updated PR, thanks!

vyasgun replied:
Updated output:
```
gvyas-mac:~ gvyas$ kubectl get pvc
NAME       STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS          AGE
my-pvc-2   Bound    pvc-1612bdc6-ce6d-40f3-beb4-d68f57eb6a34   10Gi       RWO            topolvm-provisioner   55m

gvyas-mac:~ gvyas$ crc status
CRC VM:                  Running
MicroShift:              Running (v4.15.12)
RAM Usage:               1.62GB of 3.904GB
Disk Usage:              6.824GB of 20.41GB (Inside the CRC VM)
Persistent Volume Usage: 10.67GB of 16GB (Allocated)
Cache Usage:             74.36GB
Cache Directory:         /Users/gvyas/.crc/cache
```

anjannath commented:
@vyasgun one thing i forgot to mention earlier is that the response for `/status` from the daemon API also needs to be updated to include the persistent volume size

vyasgun commented:
The `/status` changes are there: https://github.com/crc-org/crc/pull/4265/files#diff-33aaddc3c8c7beec8f2bd2c494769cf4d50dafc32852da086948ad308127d19bR73 

```
gvyas-mac:~ gvyas$ curl --unix-socket /Users/gvyas/.crc/crc-http.sock http://127.0.0.1/api/status
{"CrcStatus":"Running","OpenshiftStatus":"Running","OpenshiftVersion":"4.15.12","DiskUse":5831663616,"DiskSize":24702353408,"RAMUse":2200383488,"RAMSize":3904045056,"PersistentVolumeUse":10670309377,"PersistentVolumeSize":16000000000,"Preset":"microshift"}
```
@anjannath -- let me know if anything else needs to be added!
